sdks/python/tox.ini

```diff
@@ -17,142 +17,107 @@
 
 [tox]
 # new environments will be excluded by default unless explicitly added to envlist.
-# TODO (after BEAM-3671) add lint_py3 back in.
-envlist = py27,py27gcp,py27cython,lint_py2,docs
+envlist = py27,py27-{gcp,cython2,lint},py3-lint,docs
 toxworkdir = {toxinidir}/target/.tox
 
 [pycodestyle]
 # Disable all errors and warnings except for the ones related to blank lines.
 # pylint does not check the number of blank lines.
 select = E3
 
+# Shared environment options.
+[testenv]
+# Set [] options for pip installation of apache-beam tarball.
+extras = test
+# Don't warn that these commands aren't installed.
+whitelist_externals =
+  find
+  time
+
 [testenv:py27]
-# autocomplete_test depends on nose when invoked directly.
-deps =
-  nose==1.3.7
-  grpcio-tools==1.3.5
-whitelist_externals=find
 commands =
   python --version
   pip --version
-  # Clean up all previous python generated files.
-  - find apache_beam -type f -name '*.pyc' -delete
-  pip install -e .[test]
+  {toxinidir}/run_tox_cleanup.sh
   python apache_beam/examples/complete/autocomplete_test.py
   python setup.py test
-passenv = TRAVIS*
+  {toxinidir}/run_tox_cleanup.sh
 
-[testenv:py27cython]
+# This environment will fail in Jenkins if named "py27-cython".
+[testenv:py27-cython2]
 # cython tests are only expected to work in linux (2.x and 3.x)
 # If we want to add other platforms in the future, it should be:
 # `platform = linux2|darwin|...`
 # See https://docs.python.org/2/library/sys.html#sys.platform for platform codes
 platform = linux2
-# autocomplete_test depends on nose when invoked directly.
 deps =
-  nose==1.3.7
-  grpcio-tools==1.3.5
-  cython==0.25.2
-whitelist_externals=
-  find
-  time
+  cython==0.26.1
```

Comment:
Curious why this version? Latest is 0.28.1.

Reply:
This is the minimum version required by `setup.py` (`REQUIRED_CYTHON_VERSION`). I didn't want to increase the version beyond that and risk breaking things in this PR.

Side note: `sdks/python/container/Dockerfile` specifies `0.27.2`. *shrug*
I wish all these pinned versions were synced to a single requirements.txt file. (see [pip-compile](https://pypi.python.org/pypi/pip-tools))

Reply:
+1 on a single __requirement.txt__ file.

Reply:
That is fair to not change the version in this PR. We should consider upgrading it.

It would be nice to have a single source of truth. Although I do not see how pip-compile can update the Dockerfile. It is worth filing a JIRA for that.

Reply:
pip-compile can generate requirements.txt from setup.py, and the docker image can pull that in.
The only problem is that we have a different set of requirements depending on the extras specified (such as [test] and [gcp]), or whether cython is used.
So the solution is probably to use a script to generate a set of requirements files, using pip-compile output as an intermediate step. setup.py will be used as the single source of truth for version pinning.